lib/date_range.rb

```diff
@@ -0,0 +1,35 @@
+module Hotel
+  class DateRange
+    attr_reader :start_date, :end_date 
+    attr_accessor :dates
+
+    def initialize(start_date:, end_date:)
+      @start_date = start_date
+      @end_date = end_date
+      @dates = Array(@start_date .. @end_date)
+    
+      # User: I want an exception raised when an invalid date range is provided, so that I can't make a reservation for an invalid date range
+      if @end_date < @start_date || @start_date == @end_date
+        raise ArgumentError.new("Your check-out date must be after your check-in.")
+      end
+    end
+
+    def overlap?(daterange_instance)
+      overlap = !(daterange_instance.start_date >= end_date || daterange_instance.end_date <= start_date)
+      return overlap
```

Comment:
You don't have to assign this to a variable before you return it:

```suggestion
      return !(daterange_instance.start_date >= end_date || daterange_instance.end_date <= start_date)
```

Also, you can simplify/clarify this using [DeMorgan's Law](https://en.wikipedia.org/wiki/De_Morgan%27s_laws) and reversing the comparisons:

```suggestion
      return daterange_instance.start_date < end_date  && daterange_instance.end_date > start_date
```